src/main/java/uk/gov/pay/products/service/PaymentCreator.java

```diff
@@ -74,12 +78,38 @@ public Payment doCreate(String productExternalId) {
             paymentEntity.setExternalId(randomUuid());
             paymentEntity.setProductEntity(productEntity);
             paymentEntity.setStatus(PaymentStatus.CREATED);
-            paymentDao.persist(paymentEntity);
+            paymentEntity.setGatewayAccountId(productEntity.getGatewayAccountId());
+            paymentEntity.setReferenceNumber(randomUserFriendlyReference());
 
-            return paymentEntity;
+            AtomicInteger counter = new AtomicInteger(0);
+            return mergePaymentEntityWithReferenceNumberCheck(paymentEntity, counter);
         };
     }
 
+    private PaymentEntity mergePaymentEntityWithReferenceNumberCheck(PaymentEntity paymentEntity, AtomicInteger counter) {
+        int count = counter.getAndIncrement();
+        if(count == MAX_NUMBER_OF_RETRY_FOR_UNIQUE_REF_NUMBER) {
+            RuntimeException runtimeException = new RuntimeException("We have run out of random reference numbers");
```

Comment:
Rather than this how about saying
`Too many conflicts generating unique user friendly reference numbers for gateway account {gaid}`